util/cpuprofile/cpuprofile.go

```diff
@@ -0,0 +1,216 @@
+func Register(ch ProfileConsumer) {
+	globalCPUProfiler.register(ch)
+}
+
+// Unregister unregister a ProfileConsumer from the global CPU profiler.
+// The unregistered ProfileConsumer won't receive the cpu profile data any more.
+// This function is thread-safe.
+func Unregister(ch ProfileConsumer) {
+	globalCPUProfiler.unregister(ch)
+}
+
+// parallelCPUProfiler is a cpu profiler.
+// With parallelCPUProfiler, it is possible to have multiple profile consumer at the same time.
+// WARN: Only one running parallelCPUProfiler is allowed in the process, otherwise some profiler may profiling fail.
+type parallelCPUProfiler struct {
+	sync.Mutex
+	cs             map[ProfileConsumer]struct{}
+	notifyRegister chan struct{}
+
+	lastDataSize int
+
+	started bool
+	ctx     context.Context
+	cancel  context.CancelFunc
+	wg      sync.WaitGroup
+}
+
+// newParallelCPUProfiler crate a new parallelCPUProfiler.
+func newParallelCPUProfiler() *parallelCPUProfiler {
+	return &parallelCPUProfiler{
+		cs:             make(map[ProfileConsumer]struct{}),
+		notifyRegister: make(chan struct{}),
+	}
+}
+
+var (
+	errProfilerAlreadyStarted = errors.New("parallelCPUProfiler is already started")
+)
+
+func (p *parallelCPUProfiler) start() error {
+	p.Lock()
+	if p.started {
+		p.Unlock()
+		return errProfilerAlreadyStarted
+	}
+
+	p.started = true
+	p.ctx, p.cancel = context.WithCancel(context.Background())
+	p.Unlock()
+	p.wg.Add(1)
+	go util.WithRecovery(p.profilingLoop, nil)
+
+	logutil.BgLogger().Info("parallel cpu profiler started")
+	return nil
+}
+
+func (p *parallelCPUProfiler) stop() {
+	p.Lock()
+	if !p.started {
+		p.Unlock()
+		return
+	}
+	p.started = false
+	if p.cancel != nil {
+		p.cancel()
+	}
+	p.Unlock()
+
+	p.wg.Wait()
+	logutil.BgLogger().Info("parallel cpu profiler stopped")
+}
+
+func (p *parallelCPUProfiler) register(ch ProfileConsumer) {
+	if ch == nil {
+		return
+	}
+	p.Lock()
+	p.cs[ch] = struct{}{}
+	p.Unlock()
+
+	select {
+	case p.notifyRegister <- struct{}{}:
+	default:
+	}
+}
+
+func (p *parallelCPUProfiler) unregister(ch ProfileConsumer) {
+	if ch == nil {
+		return
+	}
+	p.Lock()
+	delete(p.cs, ch)
+	p.Unlock()
+}
+
+func (p *parallelCPUProfiler) profilingLoop() {
+	checkTicker := time.NewTicker(defProfileDuration)
+	defer func() {
+		checkTicker.Stop()
+		p.wg.Done()
+	}()
+	for {
+		select {
+		case <-p.ctx.Done():
+			return
+		case <-p.notifyRegister:
+			p.doProfiling()
+		case <-checkTicker.C:
+			p.doProfiling()
+		}
+	}
+}
+
+func (p *parallelCPUProfiler) doProfiling() {
+	if p.consumersCount() == 0 {
+		return
+	}
+
+	data := &ProfileData{Data: bytes.NewBuffer(make([]byte, 0, p.lastDataSize)), Begin: time.Now()}
+	err := pprof.StartCPUProfile(data.Data)
+	if err != nil {
+		data.Error = err
+		// notify error as soon as possible
+		p.sendToConsumers(data)
+		return
+	}
+
+	// wait 1 second
+	select {
+	case <-p.ctx.Done():
+	case <-time.After(defProfileDuration):
+	}
+
+	pprof.StopCPUProfile()
+	p.lastDataSize = data.Data.Len()
```

Comment:
how about simply allocating a fixed size buffer by default? For now you are updating the `lastDataSize` for each time, but the size may vary, sometimes big and sometimes small. I'm afraid this might not be able to avoid all re-allocations.

Reply:
I'm not sure whether a sync pool is a better choice if you really like to reduce allocations here.

Reply:
The problem is, the profile data size may vary in different kinds of loads, so It is hard to find a suitable `fixed size`.

But if the loads are same, then the profile data size of each second should also similar. And we can assume the load per second is similar to the previous second.

Reply:
How about aligning with 4KB boundaries? e.g. always to be (4 * N) KB. In this way, small workload varies will still be suitable while also possible to adjust dynamically when the workload changes drastically.

Reply:
ok